common/src/stack/command/stack/commands/set/host/power/__init__.py

```diff
@@ -25,79 +27,106 @@ class Command(stack.commands.set.host.command):
 	</param>
 
 	<param type='boolean' name='debug' optional='1'>
-	Print debug output from the ipmitool command.
+	Print debug output from the command. For ipmi capable hosts, prints
+	the output from ipmitool.
+	</param>
+
+	<param type='string' name='force' optional='1'>
+	Force a method for setting power on a host.
 	</param>
 
 	<example cmd='set host power stacki-1-1 command=reset'>
 	Performs a hard power reset on host stacki-1-1.
 	</example>
-	"""
-
-	def doPower(self, host, ipmi_ip, cmd):
-		import subprocess
-		import shlex
-
-		if not ipmi_ip:
-			return
-
-		username = self.getHostAttr(host, 'ipmi_username')
-		if not username:
-			username = 'root'
-
-		password = self.getHostAttr(host, 'ipmi_password')
-		if not password:
-			password = 'admin'
-
-		ipmi = 'ipmitool -I lanplus -H %s -U %s -P %s chassis power %s' \
-			% (ipmi_ip, username, password, cmd)
-
-		p = subprocess.Popen(shlex.split(ipmi), stdout = subprocess.PIPE,
-			stderr = subprocess.STDOUT)
-		out, err = p.communicate()
-
-		if self.debug:
-			self.beginOutput()
-			self.addOutput(host, out.decode())
-			self.endOutput(padChar='', trimOwner=True)
 
+	<example cmd='set host power stacki-1-1 command=off force=ssh'>
+	Turns off host stacki-1-1 using ssh.
+	</example>
+	"""
+	def mq_publish(self, host, cmd):
 		ttl = 60*10
 		if cmd == 'off':
 			ttl = -1
 
-		msg = { 'source' : host, 
-			'channel': 'health', 
+		msg = { 'source' : host,
+			'channel': 'health',
 			'ttl'    : ttl,
 			'payload': '{"state": "power %s"}' % cmd }
 
 		tx = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
-		tx.sendto(json.dumps(msg).encode(), 
+		tx.sendto(json.dumps(msg).encode(),
 			  ('127.0.0.1', stack.mq.ports.publish))
 		tx.close()
 
-
-
 	def run(self, params, args):
 		if not len(args):
 			raise ArgRequired(self, 'host')
 
-		cmd, debug = self.fillParams([ ('command', None, True), ('debug', False) ])
-
-		if cmd == 'status':
-			#
-			# used by "stack list host power" -- this is an easy way in which to
-			# share code between the two commands
-			#
-			# set 'debug' to True in order to get output from the status command
-			#
-			debug = True
-		elif cmd not in [ 'on', 'off', 'reset' ]:
-			raise ParamError(self, 'command', 'must be "on", "off" or "reset"')
-		
+		cmd, debug, force_imp = self.fillParams([
+			('command', None, True),
+			('debug', False),
+			('force', None)
+		])
+		power_imp = []
+		if cmd not in [ 'on', 'off', 'reset', 'status' ]:
+			raise ParamError(self, 'command', 'must be "on", "off", "reset", or "status"')
+
+		# Get all the set power implemenations
+		# Besides ipmi and ssh
+		power_imp = [imp for imp in glob(f'{Path(__file__).parent}/imp_*.py') if 'ssh' not in imp and 'ipmi' not in imp]
+		imp_names = re.findall('imp_(.*).py', '\n'.join(power_imp))
```

Comment:
The logic below could, I believe, be simplified a lot if `ipmi` and `ssh` were sorted to the front of this list. I.E. `imp_names = ["ipmi", "ssh", *re.findall('imp_(.*).py', '\n'.join(power_imp))]`. This would allow removal of the "try ssh" code path after the `for imp in imps` loop.

Reply:
I would want ssh to be the last implementation to be used (unless it was set by a flag) so I could append ipmi to the front of the list and ssh to the end.